выполнены домашние задания 8-1, 8-2, 8-3

wrhmmstr opened the pull request.

Commit:
feat: добавлен объект Message

Добавлен объект Message хранящий в себе уровень серьезности и сообщение, в рамках домашнего задания Тренировка #08-1: Обсеспечение стабильности API при расширении модели данных путем упаковки сообщения в единую структуру VO/DTO

Commit:
refactor: перевёл на использование объекта Message

Перевёл на использование объекта Message, в рамках домашнего задания Тренировка #08-1: Обсеспечение стабильности API при расширении модели данных путем упаковки сообщения в единую структуру VO/DTO

Commit:
test: обновил тестовый пример

обновил тестовый пример

Commit:
test: обновил тестовый пример

Commit:
test: обновил тестовый пример

Commit:
fix: исправил ошибку distinct вывода

Commit:
refactor: перевёл декоратор на объект Message

Перевёл декоратор на использование объекта Message в рамках домашнего задания Тренировка #08-1: Обсеспечение стабильности API при расширении модели данных путем упаковки сообщения в единую структуру VO/DTO

Commit:
feat: добавлены интерфейсы

Добавлены интерфейсы в рамках домашнего задания Тренировка #08-2: Обеспечение гибкости и расширяемости системы с полиморфизмом

Commit:
refactor: реализованы интерфейсы

Реализованы в классах с бизнес-логикой добавленные интерфейсы в рамках домашнгего задания Тренировка #08-2: Обеспечение гибкости и расширяемости системы с полиморфизмом

Commit:
refactor: ушёл от использования SeverityDecorator

Ушёл от использования SeverityDecorator, теперь в DecoratingMessageService получаю значение Severity через геттер из Severity

Commit:
refactor: ушёл от использования PagingDecorator

Ушёл от использования PagingDecorator, теперь в DecoratingMessageService получаю разделитель страницы вызов метода messageToPage из класса TimestampPageMessageDecorator через интерфейс MessageDecorator

Commit:
refactor: вынес "сборку" поведения системы в тест

Вынес "сборку" поведения системы в тест в рамках домашнего задания Тренировка #08-3 [optional]: Обеспечение большей гибкости и расширяемости системы с примемом DI

arsen-sepikhanov commented on src/main/java/com/tcs/edu/MessageDecorator.java:
```diff
@@ -0,0 +1,9 @@
+public interface MessageDecorator {
+    String decorate(Message message);
+
+    String messageToPage(int messageCount);
```
По сути этот метод используется только в контексте PagingDecorator, соответственно не очень хорошо, что он протёк в общий интерфейс

wrhmmstr replied:
Исправлено в https://github.com/wrhmmstr/qa-automation-java/pull/7

arsen-sepikhanov commented on src/main/java/com/tcs/edu/decorator/Severity.java:
```diff
@@ -3,5 +3,15 @@
  */
 public enum Severity {
-    MINOR, REGULAR, MAJOR
+    MINOR("()"), REGULAR("(!)"), MAJOR("(!!!)");
```
Класс скорее принадлежит пакету domain, чем пакету decorator

wrhmmstr replied:
Исправлено в https://github.com/wrhmmstr/qa-automation-java/pull/7

arsen-sepikhanov commented on src/test/java/com/tcs/edu/Application.java:
```diff
@@ -1,25 +1,32 @@
 
+        MessageService service = new DecoratingMessageService(
+                new ConsolePrinter(),
+                new TimestampPageMessageDecorator(),
```
Два одинаковых декоратора? Похоже на баг

wrhmmstr replied:
Я сложил TimestampDecorator и PagingDecorator в один класс потому, что не разобрался как их реализовать через один интерфейс, когда на вход ожидаются разные типы. Исправлено в https://github.com/wrhmmstr/qa-automation-java/pull/7

arsen-sepikhanov commented on src/test/java/com/tcs/edu/Application.java:
```diff
@@ -1,25 +1,32 @@
+                new ConsolePrinter(),
+                new TimestampPageMessageDecorator(),
+                new TimestampPageMessageDecorator()
+        );
```
В целом идею ты уловил правильно, но я предлагаю поднять ООПшность кода на 900%.

MessageService -  входная точка нашего API. Именно инстанс MessageService мы настраиваем, чтобы добиться нужного поведения вывода сообщений. Мне видится, что в конструкторе messageService необходимо передавать: Printer (ответственен за печатание готовых сообщений), а также vararg Decorator... decorators - список декораторов, который мы будем применять на сообщения. 

Возможные декораторы:
- TimestampDecorator - ну тут все понятно
- SeverityDecorator - тут в целом тоже
- CountingPagingDecorator - тут поинтереснее. Этот декоратор должен быть ответственным за счетчик сообщений. А также он должен прицеплять к message страницу `---` при достижении определенного условия. Также предлагаю передавать ему в конструктор int pageSize - размер страницы. А с помощью паттерна "телескопический конструктор" задать дефолтное значение pageSize = 2

wrhmmstr replied:
Будет исправлено в https://github.com/wrhmmstr/qa-automation-java/pull/8

arsen-sepikhanov commented on src/test/java/com/tcs/edu/Application.java:
```diff
@@ -1,25 +1,32 @@
+//        processMessages(message1, message2, message3, message4, message5, message6);
+//        processMessages(MessageOrder.ASC, message1, message2, message3, message4, message5, message6);
+        service.processMessages(MessageOrder.ASC, Doubling.DISTINCT, message1, message2, message3, message4, message5, message6);
+        service.processMessages(MessageOrder.DESC, Doubling.DISTINCT, message1, message2, message3, message4, message5, message6);
```
У тебя в программе сейчас есть два действия, осуществляемых над всем списком сообщений. Это устранение/неустранение дубликатов и прямой/обратный порядок. Тут сам подумай, как с помощью абстракций можно применение этой логики унифицировать

wrhmmstr replied:
Пока не понимаю сути замечания

Commit:
refactor: переместил Severity в domain

Commit:
refactor: вернул ответственность PagingDecorator

Вернул ответственность PagingDecorator, чтобы избавиться от двойной подачи на вход TimestampPageMessageDecorator
